File: lee/src/main/java/org/example/lee/题目/字符串/b_轮转数组.java
package org.example.lee.题目.字符串;

public class b_轮转数组 {

	/**
	 * rotate 基于客观事实  数组往右移动n个位置 = 先整体翻转+0-n反转+n-len反转
	 *
	 * @param nums nums
	 * @param k    k
	 */
	public void rotate(int[] nums, int k) {
		k=k%nums.length;
		toolToTurnAround(nums,0,nums.length-1);
		toolToTurnAround(nums,0,k-1);
		toolToTurnAround(nums,k,nums.length-1);
	}
	
	void toolToTurnAround(int[] nums,int start,int end){
		while (start < end){
			int h = nums[start];
			nums[start] = nums[end];
			nums[end] = h;
			start++;
			end--;
		}
	}
}

File: lee/src/main/java/org/example/lee/题目/字符串/b_轮转数组.go
package zfc

func rotate(nums []int, k int) {
	k = k % len(nums)
	reverse(nums, 0, len(nums))
	reverse(nums, 0, k)
	reverse(nums, k, len(nums))
}

func reverse(nums []int, i, j int) {
	j--
	for i < j {
		a := nums[i]
		nums[i] = nums[j]
		nums[j] = a
		i++
		j--
	}
}
